### ISSUE TITLE
Location of included launch files cannot be determined

### ISSUE THREAD

**Issue by dseredyn** [ros2/launch#799]
## Bug report

- Operating System:
  - Ubuntu 22.04
- Installation type:
  binary
- Version or commit hash:
  humble, from distro
- DDS implementation:
  - 
- Client library (if applicable):
  - 

#### Steps to reproduce issue
I'm trying to make a tool for examination of launch file structure and to visualize a tree of included launch files (python).

The following code, where ent is IncludeLaunchDescription object, should output a path to the included launch file, but instead it returns a string that refers to Python object pointer, e.g. "<launch.substitutions.path_join_substitution.PathJoinSubstitution object at 0x74b241481780>"

if isinstance(ent, IncludeLaunchDescription):
  print(isinstance(ent.launch_description_source.location, str))
  print(ent.launch_description_source.location)

ACTUAL output:
True
<launch.substitutions.path_join_substitution.PathJoinSubstitution object at 0x74b241481780>

EXPECTED output:
True
/home/user/ws/share/...

This should be fixed to enable error reporting for individual launch files included from other launch files. Now it is impossible to automatically determine which launch file has defined a problematic action.

I'm using PAL robotics scripts to build launch files. The function that includes a launch file is:

def include_launch_py_description(
        pkg_name: SomeSubstitutionsType,
        paths: List[SomeSubstitutionsType],
        **kwargs) -> Action:
    """
    Return IncludeLaunchDescription for the file inside pkg at paths.

    Example:
    -------
        include_launch_py_description('my_pkg', ['launch', 'my_file.launch.py'])
        returns file IncludeLaunchDescription from PATH_TO_MY_PKG_SHARE/launch/my_file.launch.py

    """
    pkg_dir = FindPackageShare(pkg_name)
    full_path = PathJoinSubstitution([pkg_dir] + paths)

    return IncludeLaunchDescription(
        PythonLaunchDescriptionSource(
            full_path),
        **kwargs)

**Comment by fujitatomoya** [comment]
@wjwwood can you take a look about this?

**Comment by christophebedard** [comment]
@dseredyn sorry we haven't followed up here. What's the latest status of your launch examination tool?

tl;dr substitutions need to be performed by calling `LaunchDescriptionSource.get_launch_description()` before using `LaunchDescriptionSource.location`

----

> Also, it looks like IncludeLaunchDescription convert Substitutions into a string <launch.substitutions.path_join_substitution.PathJoinSubstitution object at ...>:

Yeah, I think this is the issue. But I think it's also intentional, according to the documentation: https://github.com/ros2/launch/blob/57c1aeec7b18be4e0e93c91a4b069d4437f5b839/launch/launch/launch_description_source.py#L95-L107

This situation happens if `LaunchDescriptionSource.get_launch_description()` hasn't been called yet. If it has been called, it returns `self.__expanded_location`, which contains the proper location with all substitutions performed. Otherwise, it seemingly intentionally returns a "list of Substitution instances converted to strings."

The actual location is not known until the substitutions are performed, which happens later at runtime, when the `IncludeLaunchDescription` action gets executed: https://github.com/ros2/launch/blob/57c1aeec7b18be4e0e93c91a4b069d4437f5b839/launch/launch/actions/include_launch_description.py#L214. Before that point, the best we can do is to return some kind of "representation" of the location.

This behaviour is common in the `launch` code, but it is a bit inconsistent. In some cases, getters for properties that include substitutions cannot be used before substitutions are performed (i.e., an error is raised). In other cases, getters intentionally expose values with substitutions, and it's the caller's responsibility to perform the substitutions. Unfortunately, this challenge also comes up repeatedly when trying to build introspection tools for `launch`.

In the case of `LaunchDescriptionSource.location`, it tries to return _something_, which might be an invalid location/path. I'm inclined to consider this "working as designed," but it would be interesting to change it so that an error is raised if someone tries to get the location before substitutions have been performed, and see if something downstream breaks. This would make this behaviour more obvious.

**Comment by wjwwood** [comment]
> In the case of LaunchDescriptionSource.location, it tries to return something, which might be an invalid location/path. I'm inclined to consider this "working as designed," but it would be interesting to change it so that an error is raised if someone tries to get the location before substitutions have been performed, and see if something downstream breaks. This would make this behaviour more obvious.

This is correct. The location is necessarily a list of substitutions, and all introspection tools will necessarily either need to actually run the launch file (sort of what we're trying to avoid) or make a "best effort" to resolve substitutions for introspection. If the substitutions are non-trivial, it will just make anything past that opaque to introspection.
